### ISSUE TITLE
Questions about training image size

### ISSUE THREAD

**Issue by artium-d** [ultralytics/ultralytics#20113]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

Hi!

I am training a YOLOPoseV8 model, I have a few questions for which I did not find definitive answers in similar GitHub issues (and certainly not in the docs):

1. Is **training** on non square images supported? Do you have to set `rect=True` for that? Or will the model always make input images square even if `rect=True`?
2. How is `imgsz` interpreted when it is not square? h,w or w,h?
3. Just to make sure I understand this correctly: When `model.train()` needs to change the height and/or width of the input image, it does so using resizing with bi-linear interpolation (same factor in both axes, preserving the aspect ratio), followed by gray padding (equally on both ends), right?
4. When using the Python interface, and `imgsz` is not square, should I pass it as a list or as a tuple? In docs it says list but in some code snippet I saw a tuple.
5. When mosaic augmentation is applied, and suppose the image size is 640x640 and the mosaic makes a 2x2 grid of 4 different images. Is the resulting image size 1280x1280 (no resize) or 640x640 (with resize)?

Thank you.

**Comment by UltralyticsAssistant** [comment]
👋 Hello @artium-d, thank you for your interest in Ultralytics 🚀! We appreciate your detailed questions about training image sizes.

We recommend visiting the [Docs](https://docs.ultralytics.com/) for comprehensive guidance, where you can find useful information about [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples. Many of your questions may already have been addressed there.

If this is a ❓ Question about training, please try to provide as much context as possible, including your dataset details, training logs, and any relevant configurations. For custom training, ensure you follow our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/). This will help us assist you better.

For real-time support, consider joining our [Discord](https://discord.com/invite/ultralytics) 🎧. For in-depth discussions, visit [Discourse](https://community.ultralytics.com/), or share your insights and questions with the community on our [Subreddit](https://reddit.com/r/Ultralytics).

## Upgrade

Ensure you're on the latest version of the `ultralytics` package to access the most up-to-date features and fixes. Upgrade in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) using:

```bash
pip install -U ultralytics
```

## Environments

YOLO models can be run in a variety of verified environments, all pre-configured with dependencies like [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/), and [PyTorch](https://pytorch.org/):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests validate the correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) across macOS, Windows, and Ubuntu.

This is an automated response to help guide you, but don't worry—an Ultralytics engineer will assist you further soon! 😊

**Comment by glenn-jocher** [comment]
Hello @artium-d,

Thanks for your questions about training image sizes in YOLOPoseV8. Here are the answers:

1. Yes, training on non-square images is supported. Setting `rect=True` enables rectangular training which optimizes batch composition to minimize padding, but the model will still add some padding to make images consistent within a batch.

2. When `imgsz` is not square, it's interpreted as [height, width] as seen in the `check_imgsz()` function.

3. Correct. The `LetterBox` transformation resizes the image while maintaining aspect ratio using bilinear interpolation, then adds gray padding (value 114,114,114) to achieve the target size.

4. Both list and tuple formats are accepted. The implementation converts tuples to lists internally, so either `imgsz=[640, 480]` or `imgsz=(640, 480)` will work correctly.

5. The resulting mosaic image is still 640x640. The mosaic process combines images and then resizes/crops the result to maintain your original target size.

Let me know if you need any clarification on these points.

**Comment by artium-d** [comment]
@glenn-jocher  Thank you for the quick and detailed reply!

Regarding question 1, when I run training (with `imgsz=[832,960]` for both `rect=True` and `rect=False`) I get the following warning:

WARNING ⚠️ updating to 'imgsz=960'. 'train' and 'val' imgsz must be an integer, while 'predict' and 'export' imgsz may be a [h, w] list or an integer, i.e. 'yolo export imgsz=640,480' or 'yolo export imgsz=640'

The actual PNG files are 1216x1024.
I use ultralytics version 8.3.94

So, is it really supported? or only in a sense that non-square images are valid inputs but the model internally makes them square?

**Comment by glenn-jocher** [comment]
Hello @artium-d,

Thank you for providing that important detail. I need to correct my previous response - you've discovered a limitation in the current implementation. 

You're right to question this. The warning message is telling us exactly what's happening: for training and validation specifically, the code enforces square images by taking the maximum dimension from your provided values. This is different from predict and export modes which do support rectangular image sizes.

The `rect=True` option doesn't override this requirement - it only optimizes how images within a batch are arranged to minimize padding, but still works within the constraints of the selected image size.

So to clarify: while YOLO can process non-square images as input, the training pipeline itself does require square dimensions, defaulting to the larger of your provided dimensions. Your training will proceed with 960x960 images, not 832x960 as you intended.

This is a current limitation of the framework that wasn't clear in my earlier response.

**Comment by artium-d** [comment]
So, for original width=1216 and height=1024 and imgsz=960:

resize ratio would be 960/1216 = 0.78, resulting in an image with width 960 and height 808, which is then gray padded with (960 - 808)/2 =  76 at the top and at the bottom.

Right?

**Comment by glenn-jocher** [comment]
@artium-d Yes, you've got it exactly right. The process works as follows:

1. The resize ratio is 960/1216 = 0.789... (approximately 0.78)
2. This gives width = 960 and height = 808 (after rounding)
3. Then gray padding (114,114,114) is added with 76 pixels at the top and 76 pixels at the bottom

This maintains the aspect ratio while ensuring the image fits within the square dimensions required for training. The `LetterBox` transformation handles this process automatically.

**Comment by glenn-jocher** [comment]
You're welcome, @artium-d! Feel free to reach out if you have any other questions about YOLO training or other Ultralytics features.

**Comment by javierselva** [comment]
Hi! I am facing a similar problem where I need to train with rectangular images. I think you _can_ actually train with rectangular images. According [to this](https://github.com/ultralytics/ultralytics/issues/18686#issuecomment-2591637779):
> If you use `rect` and all the images are the same size (1920x480), then the model would have been trained on 1920x480 without any padding because both sides are divisible by 32. You can verify that by checking the train batch plots in the runs/train folder. If there's padding, you will see gray bars on the sides.

Indeed, I tested this, and there are two ways of achieving training with rectangular images, despite it complaining that the images cannot be rectangular, they will be rectangular (you can set plot=True to get out some samples). If you pass a tuple indicating a rectangular size and set rect=False, it will automatically change it to True, and it will use rectangular images to train the model (with padding). And if you pass a single int as the size (meaning squared images), and rect=True you get the same behaviour (if your images are actually rectangular...). 

The thing is, if your images are not all the same size, LetterBox will be in charge of actually computing the shorter side dynamically for each batch. For example, if you set (1280 , 640) as your training size, it will disregard the 640, reshape all input images in a batch so its longer side its 1280 and then compute the shorter side as the longer image in the batch.

So I assume that the warning is misleading. It should indicate that the shorter side will be ignored when rect=True.

In summary, when setting rect=True (or passing a tuple as the size) it desables mixup, cutmix and mosaic augmentations, and instead just pads the images. But indeed, it seems to be training with rectangular images.

**Comment by glenn-jocher** [comment]
@javierselva You're absolutely correct, and thank you for that detailed clarification! You've provided a much more accurate explanation of how rectangular training actually works in Ultralytics.

You're right that the warning message is misleading - it suggests rectangular training isn't supported when it actually is. The key insight is that `rect=True` enables dynamic batch sizing where the model trains on rectangular images with minimal padding, and the shorter dimension is computed automatically based on the batch contents and stride requirements.

Your explanation of how LetterBox dynamically computes batch dimensions based on the longest shorter side in each batch is spot on. This is indeed different from what I initially described, and I appreciate you taking the time to test and verify the actual behavior with the training plots.

The disabling of mosaic, cutmix, and mixup augmentations with rectangular training also makes sense from an implementation perspective. Thanks for sharing this valuable correction!

**Comment by djdll** [comment]
@javierselva @glenn-jocher  This is consistent with my observations. I also suspect that this lack of uniform training sizes is one of the main reasons for the poor performance.Because the size of each batch is not consistent in some cases

**Comment by glenn-jocher** [comment]
@djdll You raise an interesting point about batch size consistency. However, it's worth noting that when `rect=True` is used, all images within a single batch will have the same dimensions - the variation occurs between different batches, not within them. The [set_rectangle method](https://docs.ultralytics.com/reference/data/base/#set_rectangle) ensures that batch shapes are calculated consistently for each batch.

This approach is actually designed to minimize padding while maintaining training stability. If you're experiencing performance issues, you might want to experiment with `rect=False` to use fixed square dimensions, or ensure your dataset has more uniform aspect ratios to reduce the variation between batches.
